odl/util/numerics.py

```diff
@@ -836,6 +838,165 @@ def zscore(arr):
+    - `slice` objects,
+    - index arrays, i.e., 1D array-like objects containing integers,
+
+    and combinations of the above. It does not work with boolean arrays
+    or more advanced "fancy indexing".
+
+    Parameters
+    ----------
+    shape : sequence of ints
+        Number of entries per axis in the "simulated" array.
+    indices : index expression
+        Single instance or sequence of integer, `slice` and/or `array-like`
+        objects that should be used to simulate indexing of an array.
+        If index arrays are used, they must all be one-dimensional and have
+        the same length.
+
+    Returns
+    -------
+    sliced_shape : tuple of ints
+        The shape after simulated indexing.
+    collapsed_axes : tuple of ints
+        The axes that will be gone after indexing.
+    leftmost_index_array_axis : int or None
+        Index of the leftmost index array in the provided ``indices``,
+        or ``None`` if there is no index array.
+
+    Raises
+    ------
+    TypeError
+        If an invalid type of indexing object is used. This can be caught
+        by upstream code to switch to lazy indexing.
+
+    Examples
+    --------
+    A single integer slices along the first axis (index does not
+    matter as long as it lies within the bounds):
+
+    >>> shape = (3, 4, 5, 6)
+    >>> simulate_slicing(shape, 0)  # arr[0]
+    ((4, 5, 6), (0,), None)
+    >>> rn[2]
+    ((4, 5, 6), (0,), None)
+
+    Multiple indices slice into corresponding axes from the left:
+
+    >>> simulate_slicing(shape, (0, 0))  # arr[0, 0]
+    ((5, 6), (0, 1), None)
+    >>> simulate_slicing(shape, (0, 1, 1))  # arr[0, 1, 1]
+    ((6,), (0, 1, 2), None)
+
+    Ellipsis (``...``) and ``slice(None)`` (``:``) can be used to keep
+    one or several axes intact:
+
+    >>> # arr[0, :, 1, :]
+    >>> simulate_slicing(shape, (0, slice(None), 1, slice(None)))
+    ((4, 6), (0, 2), None)
+    >>> simulate_slicing(shape, (Ellipsis, 0, 0))  # arr[..., 0, 0]
+    ((3, 4), (2, 3), None)
+
+    With slices, parts of axes can be selected:
+
+    >>> # arr[0, :3, 1:4, ::2]
+    >>> simulate_slicing(
+    ...     shape, (0, slice(None, 3), slice(1, 4), slice(None, None, 2)))
+    ((3, 3, 3), (0,), None)
+
+    Array-like objects (must all have the same 1D shape) of integers are
+    treated as follows: if their common length is ``n``, a new axis
+    of length ``n`` is created at the position of the leftmost index
+    array, and all index array axes are collapsed (Note: this is
+    not so useful for spaces, more so for elements):
+
+    >>> # arr[0, 0, [0, 1, 0, 2], :]
+    >>> simulate_slicing(shape, (0, 0, [0, 1, 0, 2], slice(None)))
+    ((4, 6), (0, 1), 2)
+    >>> arr[:, [1, 1], [0, 2], :]
+    >>> simulate_slicing(shape, (slice(None), [1, 1], [0, 2], slice(None)))
+    ((3, 2, 6), (2,), 1)
+    >>> # arr[[2, 0], [3, 3], [0, 1], [5, 2]]
+    >>> simulate_slicing(shape, ([2, 0], [3, 3], [0, 1], [5, 2]))
+    ((2,), (1, 2, 3), 0)
+    """
```

Comment:
Show usage with `np.s_` which makes stuff like `np.s_[..., 2]` much easier to read

Reply:
Bump on this

Reply:
I thought I'd done this ...